packages/transaction-controller/src/TransactionController.ts

```diff
@@ -666,63 +897,82 @@ export class TransactionController extends BaseControllerV1<
-    const transactionMeta: TransactionMeta = existingTransactionMeta || {
-      // Add actionId to txMeta to check if same actionId is seen again
-      actionId,
-      chainId,
-      dappSuggestedGasFees,
-      deviceConfirmedOn,
-      id: random(),
-      origin,
-      securityAlertResponse,
-      status: TransactionStatus.unapproved as TransactionStatus.unapproved,
-      time: Date.now(),
-      txParams,
-      userEditedGasLimit: false,
-      verifiedOnBlockchain: false,
-      type: transactionType,
-    };
-
-    await this.updateGasProperties(transactionMeta);
+    let addedTransactionMeta: TransactionMeta = existingTransactionMeta
+      ? { ...existingTransactionMeta }
+      : {
+          // Add actionId to txMeta to check if same actionId is seen again
+          actionId,
+          chainId,
+          dappSuggestedGasFees,
+          deviceConfirmedOn,
+          id: random(),
+          origin,
+          securityAlertResponse,
+          status: TransactionStatus.unapproved as TransactionStatus.unapproved,
+          time: Date.now(),
+          txParams,
+          userEditedGasLimit: false,
+          verifiedOnBlockchain: false,
+          type: transactionType,
+        };
+
+    await this.updateGasProperties(addedTransactionMeta);
 
     // Checks if a transaction already exists with a given actionId
     if (!existingTransactionMeta) {
       // Set security provider response
       if (method && this.securityProviderRequest) {
         const securityProviderResponse = await this.securityProviderRequest(
-          transactionMeta,
+          addedTransactionMeta,
           method,
         );
-        transactionMeta.securityProviderResponse = securityProviderResponse;
+        addedTransactionMeta.securityProviderResponse =
+          securityProviderResponse;
       }
 
       if (!this.isSendFlowHistoryDisabled) {
-        transactionMeta.sendFlowHistory = sendFlowHistory ?? [];
+        addedTransactionMeta.sendFlowHistory = sendFlowHistory ?? [];
       }
       // Initial history push
       if (!this.isHistoryDisabled) {
-        addInitialHistorySnapshot(transactionMeta);
+        addedTransactionMeta = addInitialHistorySnapshot(addedTransactionMeta);
       }
 
-      await updateSwapsTransaction(transactionMeta, transactionType, swaps, {
-        isSwapsDisabled: this.isSwapsDisabled,
-        cancelTransaction: this.cancelTransaction.bind(this),
-        // TODO: Replace `any` with type
-        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        controllerHubEmitter: this.hub.emit.bind(this.hub) as any,
-      });
+      addedTransactionMeta = updateSwapsTransaction(
+        addedTransactionMeta,
+        transactionType,
+        swaps,
+        {
+          isSwapsDisabled: this.isSwapsDisabled,
+          cancelTransaction: this.cancelTransaction.bind(this),
+        },
+      );
+
+      if (transactionType === TransactionType.swapApproval) {
```

Comment:
The knowledge of the alternate swap transaction types and the resulting events was intentionally encapsulated in the util since it's an extension of the controller in this instance.

Could we just provide the messaging system instance rather than breaking encapsulation?

Reply:
Sure, I've undone these changes to do what you requested.